src/lib/x509/certstor_system_windows/certstor_windows.cpp

```diff
@@ -0,0 +1,236 @@
+
+std::vector<X509_DN> Certificate_Store_Windows::all_subjects() const
+   {
+   std::vector<X509_DN> subject_dns;
+   for(auto& store_name : ::cert_store_names)
+      {
+      Handle_Guard<HCERTSTORE> windows_cert_store = openCertStore(store_name.c_str());
+      Handle_Guard<PCCERT_CONTEXT> cert_context = nullptr;
+
+      // Handle_Guard::assign exchanges the underlying pointer. No RAII is needed here, because the Windows API takes care of
+      // freeing the previous context.
+      while(cert_context.assign(CertEnumCertificatesInStore(windows_cert_store.get(), cert_context.get())))
+         {
+         X509_Certificate cert(cert_context->pbCertEncoded, cert_context->cbCertEncoded);
+         subject_dns.push_back(cert.subject_dn());
+         }
+      }
+
+   return subject_dns;
+   }
+
+std::shared_ptr<const X509_Certificate>
+Certificate_Store_Windows::find_cert(const Botan::X509_DN&           subject_dn,
+                                     const std::vector<uint8_t>& key_id) const
+   {
+   const auto certs = find_all_certs(subject_dn, key_id);
+   return certs.empty() ? nullptr : certs.front();
+   }
+
+std::vector<std::shared_ptr<const X509_Certificate>> Certificate_Store_Windows::find_all_certs(
+         const X509_DN& subject_dn,
+         const std::vector<uint8_t>& key_id) const
+   {
+   std::vector<uint8_t> dn_data;
+   DER_Encoder encoder(dn_data);
+   subject_dn.encode_into(encoder);
+
+   CERT_NAME_BLOB blob;
+   blob.cbData = static_cast<DWORD>(dn_data.size());
+   blob.pbData = reinterpret_cast<BYTE*>(dn_data.data());
+
+   std::vector<std::shared_ptr<const X509_Certificate>> certs;
+   for(auto& store_name : ::cert_store_names)
+      {
+      Handle_Guard<HCERTSTORE> windows_cert_store = openCertStore(store_name);
+      Handle_Guard<PCCERT_CONTEXT> cert_context = nullptr;
+      while(cert_context.assign(CertFindCertificateInStore(
+                                   windows_cert_store.get(), PKCS_7_ASN_ENCODING | X509_ASN_ENCODING,
+                                   NULL, CERT_FIND_SUBJECT_NAME,
+                                   &blob, cert_context.get())))
+         {
+         auto cert = std::make_shared<X509_Certificate>(cert_context->pbCertEncoded, cert_context->cbCertEncoded);
+         if(!already_contains_certificate(certs, cert) && (key_id.empty() || cert->subject_key_id() == key_id))
+            {
+            certs.push_back(cert);
+            }
+         }
+      }
+   return certs;
+   }
+
+std::shared_ptr<const Botan::X509_Certificate>
+Certificate_Store_Windows::find_cert_by_pubkey_sha1(
+   const std::vector<uint8_t>& key_hash) const
+   {
+   if(key_hash.size() != 20)
+      {
+      throw Invalid_Argument("Certificate_Store_Windows::find_cert_by_pubkey_sha1 invalid hash");
+      }
+
+   CRYPT_HASH_BLOB blob;
+   blob.cbData = static_cast<DWORD>(key_hash.size());
+   blob.pbData = const_cast<BYTE*>(key_hash.data());
+
+   for(auto& store_name : ::cert_store_names)
```

Comment:
Design q. This enumerates all of the cert stores. Is it in some contexts reasonable to only use some subset of them? That would argue for the class to take a list of which specific ones to use to allow the user to control this if needed.
(I don't know the answer to this, I'm not really that familiar with how Windows cert stores are used in practice.)

Reply:
s.a. It might make sense in some cases, but I don't know if there is much gain in doing that right now.

Reply:
OK. If it comes up later we can add a new constructor taking the list and leaving the default constructor with the current behavior as default.